kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/data/DefaultJsonRow.java

```diff
@@ -140,17 +145,84 @@ private static Object decodeElement(JsonNode jsonValue, DataType dataType) {
             return jsonValue.booleanValue();
         }
 
+        if (dataType instanceof ByteType) {
+            throwIfTypeMismatch(
+                "byte",
+                jsonValue.canConvertToExactIntegral() && jsonValue.intValue() <= Byte.MAX_VALUE,
+                jsonValue
+            );
+            return jsonValue.numberValue().byteValue();
+        }
+
+        if (dataType instanceof ShortType) {
+            throwIfTypeMismatch(
+                "short",
+                jsonValue.canConvertToExactIntegral() && jsonValue.intValue() <= Short.MAX_VALUE,
+                jsonValue
+            );
+            return jsonValue.numberValue().shortValue();
+        }
+
         if (dataType instanceof IntegerType) {
             throwIfTypeMismatch(
-                "integer", jsonValue.isIntegralNumber() && !jsonValue.isLong(), jsonValue);
+                "integer",
+                jsonValue.isIntegralNumber() && jsonValue.canConvertToInt(),
+                jsonValue);
             return jsonValue.intValue();
         }
 
         if (dataType instanceof LongType) {
-            throwIfTypeMismatch("long", jsonValue.isIntegralNumber(), jsonValue);
+            throwIfTypeMismatch("long",
+                jsonValue.isIntegralNumber() && jsonValue.canConvertToLong(), jsonValue);
             return jsonValue.numberValue().longValue();
         }
 
+        if (dataType instanceof FloatType) {
+            switch (jsonValue.getNodeType()) {
+                case NUMBER:
+                    throwIfTypeMismatch(
+                        "float",
+                        jsonValue.doubleValue() <= Float.MAX_VALUE,
+                        jsonValue
+                    );
+                    return jsonValue.floatValue();
+                case STRING:
+                    switch (jsonValue.asText()) {
+                        case "NaN":
+                            return Float.NaN;
+                        case "+INF": case "+Infinity": case "Infinity":
+                            return Float.POSITIVE_INFINITY;
+                        case "-INF": case "-Infinity":
+                            return Float.NEGATIVE_INFINITY;
+                    }
+                default:
+                    throwIfTypeMismatch("float", false, jsonValue);
+            }
+        }
+
+        if (dataType instanceof DoubleType) {
+            switch (jsonValue.getNodeType()) {
+                case NUMBER:
+                    throwIfTypeMismatch(
+                        "double",
+                        jsonValue.doubleValue() <= Double.MAX_VALUE,
```

Comment:
do we need this check?

Reply:
Seems like the spark parser doesn't check for overflow; I'm good to remove all the checks but I think in theory this could be a java big decimal that won't fit

Reply:
Oh I misunderstood.. I've updated these checks anyways